src/node/lattice.rs

```diff
@@ -0,0 +1,518 @@
+impl ExecutionLattice {
+    /// Creates a new instance of `ExecutionLattice`.
+    pub fn new() -> Self {
+        ExecutionLattice {
+            forest: Arc::new(Mutex::new(StableGraph::new())),
+            roots: Arc::new(Mutex::new(HashSet::new())),
+            run_queue: Arc::new(Mutex::new(VecDeque::new())),
+        }
+    }
+
+    /// Add an event to the lattice.
+    ///
+    /// This function moves the passed event into the lattice, and inserts the appropriate edges to
+    /// existing events in the graph based on the partial order defined in [`OperatorEvent`].
+    pub async fn add_event(&self, event: OperatorEvent) {
+        // Take locks over everything.
+        let mut forest = self.forest.lock().await;
+        let mut roots = self.roots.lock().await;
+        let mut run_queue = self.run_queue.lock().await;
+
+        // Iterate through all the roots, and figure out where to add a dependency edge.
+        let mut dfs = DfsPostOrder::empty(&*forest);
+        let mut dependencies: HashSet<NodeIndex<u32>> = HashSet::new();
+        let mut add_edges_to: Vec<NodeIndex<u32>> = Vec::new();
+        let mut add_edges_from: Vec<NodeIndex<u32>> = Vec::new();
+        let mut remove_roots: Vec<(usize, NodeIndex<u32>)> = Vec::new();
+
+        for root in roots.iter() {
+            // This function maintains the stack when iterating over the graph so we do not have to
+            // check the same node in the graph twice.
+            dfs.move_to(*root);
+
+            // Retrieve the next node in DFS order from the forest.
+            while let Some(nx) = dfs.next(&*forest) {
+                // Check if any of the children of the current node in DFS traversal already have a
+                // dependency on the node we are trying to add.
+                let mut dependent_child: bool = false;
+                for child in forest.neighbors(nx) {
+                    if dependencies.contains(&child) {
+                        dependencies.insert(nx);
+                        dependent_child = true;
+                        break;
+                    }
+                }
+
+                // If the event we are trying to add has already been added as a dependency of one
+                // of the children of the DFS node, then move to the next root of the forest.
+                if dependent_child {
+                    break;
+                }
+
+                if let Some(node) = forest.node_weight(nx).unwrap().as_ref() {
+                    // If the event we are trying to add is dependent on the DFS node, then
+                    // add an edge from the DFS node to the event to be added.
+                    if &event <= node {
+                        // Check if any of the children of the DFS node should be dependent on the
+                        // event we are trying to add.
+                        let mut remove_edges: Vec<NodeIndex<u32>> = Vec::new();
+                        for child in forest.neighbors(nx) {
```

Comment:
Do you have to add these dependencies if &event < node? Isn't enough to add them when &event == node?

Reply:
I think it was mainly for cases when events get reordered. For example, you have a watermark for time `t2` being delivered before watermark for time `t1`. (where `t1` < `t2`) Then, the no dependencies check above would fail, but they would still not satisfy `==`.